File: exercises/domain-1/block-1.4/solution/solution.py
import json
import os
from datetime import datetime, timezone
import asyncio


class ResumablePipeline:
    def __init__(self, state_file: str):
        self.state_file = state_file
        self.steps = []

    def add_step(self, name, fn):
        self.steps.append((name, fn))
        return self

    def _load_state(self):
        try:
            if os.path.exists(self.state_file):
                with open(self.state_file) as f:
                    return json.load(f)
        except Exception:
            pass
        return {"completed_steps": [], "last_updated": datetime.now(timezone.utc).isoformat()}

    def _save_state(self, state):
        tmp = self.state_file + ".tmp"
        with open(tmp, "w") as f:
            json.dump(state, f, indent=2)
        os.replace(tmp, self.state_file)

    async def run(self):
        state = self._load_state()

        for name, fn in self.steps:
            if name in state["completed_steps"]:
                print(f"  ⏭ Skipped: {name}")
                continue

            await fn(name)

            state["completed_steps"].append(name)
            state["last_updated"] = datetime.now(timezone.utc).isoformat()
            self._save_state(state)


STATE_FILE = "/tmp/pipeline_state.json"
if os.path.exists(STATE_FILE):
    os.remove(STATE_FILE)

pipeline = ResumablePipeline(STATE_FILE)


async def step_fetch(name): print(f"  ✓ {name}: fetched")
async def step_transform(name):
    print(f"  ✓ {name}: transformed")
    raise RuntimeError("Interrupted!")
async def step_save(name): print(f"  ✓ {name}: saved")


pipeline.add_step("step-1-fetch", step_fetch)
pipeline.add_step("step-2-transform", step_transform)
pipeline.add_step("step-3-save", step_save)


async def main():
    print("=== Run 1 ===")
    try:
        await pipeline.run()
    except Exception:
        print("Interrupted.\n=== Run 2 ===")
        await pipeline.run()
        print("Done!")


asyncio.run(main())

File: exercises/domain-1/block-1.4/solution/solution.ts
import fs from "fs"

interface PipelineState {
  completedSteps: string[]
  lastUpdated: string
}

type StepFn = (name: string) => Promise<void>

class ResumablePipeline {
  private stateFile: string
  private steps: Array<{ name: string; fn: StepFn }>

  constructor(stateFile: string) {
    this.stateFile = stateFile
    this.steps = []
  }

  addStep(name: string, fn: StepFn): this {
    this.steps.push({ name, fn })
    return this
  }

  private loadState(): PipelineState {
    try {
      if (fs.existsSync(this.stateFile)) {
        return JSON.parse(fs.readFileSync(this.stateFile, "utf8"))
      }
    } catch { /* corrupt file, start fresh */ }
    return { completedSteps: [], lastUpdated: new Date().toISOString() }
  }

  private saveState(state: PipelineState): void {
    const tmp = this.stateFile + ".tmp"
    fs.writeFileSync(tmp, JSON.stringify(state, null, 2))
    fs.renameSync(tmp, this.stateFile)
  }

  async run(): Promise<void> {
    const state = this.loadState()

    for (const { name, fn } of this.steps) {
      if (state.completedSteps.includes(name)) {
        console.log(`  ⏭ Skipped (already done): ${name}`)
        continue
      }

      await fn(name)

      state.completedSteps.push(name)
      state.lastUpdated = new Date().toISOString()
      this.saveState(state)
    }
  }
}

const STATE_FILE = "/tmp/pipeline-state.json"
if (fs.existsSync(STATE_FILE)) fs.unlinkSync(STATE_FILE)

const pipeline = new ResumablePipeline(STATE_FILE)

pipeline
  .addStep("step-1-fetch", async (name) => { console.log(`  ✓ ${name}: fetched`) })
  .addStep("step-2-transform", async (name) => {
    console.log(`  ✓ ${name}: transformed`)
    throw new Error("Interrupted!")
  })
  .addStep("step-3-save", async (name) => { console.log(`  ✓ ${name}: saved`) })

console.log("=== Run 1 ===")
pipeline.run().catch(() => {
  console.log("Interrupted.\n=== Run 2 (resuming) ===")
  pipeline.run().then(() => console.log("Done!"))
})
